refactor(ui): break up and clean up `StarterSelectUiHandler`, reduce number of containers

Wlowscha opened the pull request:
## What are the changes the user will see?
Reduced memory usage from the starter select screen.

<!-- ## Changelog cutoff (DO NOT REMOVE/EDIT) -->

## Why am I making these changes?
This PR has two main goals:
1) Functionally change the starter select screen to use 81 containers instead of one per each starter,
2) Refactor and split up the various methods in it to reduce complexity and hopefully make it easier to make changes in the future.

## What are the changes from a developer perspective?

(This is a continuation of #6390, see there for older commits).

List of bugs found by @fabske0 here: https://discord.com/channels/1125469663833370665/1412372243492634666/1414529712369176607
(they should all be fixed, hopefully).

This is by necessity a gigantic PR due to the size of the files involved.

Note that, both due to the size of the files involved and their UI nature, manual testing to find bugs is most likely the best approach to review this PR.


Many functions in the starter select ui handler and adjacent files have been updated to use `StarterSpeciedId` in their arguments. This prevents passing a generic `Species` or `SpeciesId` by mistake which may cause issues. In particular, this affects all functions which interact with starter data in the game data.

Cursors for ability, nature etc. which stored a copy of the preferences for the current species have been removed as they were redundant.


### Easy to digest changes

Renamed the `StarterAttributes` type to `StarterPreferences` to better convey its usage, and renamed the `StarterPreferences` interface to `AllStarterPreferences`.

Introduced a `getSpeciesDefaultDexAttrProps` function in `game-data.ts` which returns, for a given species, default choices of preferences for shiny, gender and form based on caught data.

There are some changes to the Pokedex files, but they mostly consist of renaming variables or using some of the new utility functions.

Multiple functions and data structures have been moved from `StarterSelectUiHandler` to the new `starter-select-ui-utils.ts` file. In particular, the definition of `SpeciesDetails` is here, and is now the same in all pokedex files as well.
Language settings are also here, as well as various functions that convert `StarterPreferences` to different data formats (e.g. `SpeciesDetails` or `DexAttrProps`.
Functions such as `isPassiveAvailable` have also been moved here, although some are still wrapped in a helper method when used in the starter select ui class.

The `StarterContainer` now has a `setSpecies` method so that it can be updated (changing species) on the fly.


### Heavier changes

The `StarterSelectUiHandler` has been further split up, with many methods going into a `StarterSummary` class. The idea is for the latter to manage a single starter pokémon---essentially, everything that is shown on the left side of the screen---while `StarterSelectUiHandler` still manages the box, filters, the party tray etc. This takes out approximately 1k lines of code (still leaving almost 4k in the `starter-select-ui-handler.ts` file.

Some smaller methods, referring to specific ui elements, have been extracted from the enormous `setup` method to be a little more manageable: `setupFilterBar`, `setupPartyColumn`, and `setupInstructionButtons`, each between 100-150 lines.

The even scarier `processInput` function has also been split up into several methods with smaller scope, interacting with specific UI elements. Much of this complexity went into the `openPokemonMenu` function, which is still over 500 lines long.

The most significant change in starter select which is not only moving code around is that the number of starter containers is now fixed to 81, instead of being equal to the number of starters; this implements the same approach already used in the Pokédex.


### Below is a list of methods of `StarterSelectUiHandler` with some comments which may help while reviewing.

The functions `setup`, `setupFlterBar`, `setupPartyColumn`, `setupInstructionButtons` are about 500 total lines uniquely to set up ui elements and some other stuff like the list of starters.
These should be okay for now.

`initStarterPrefs` is a function which sanitizes the preferences for a given starter, basically ensures that no invalid values are selected as preferences, and it also applies challenges to prevent unwanted options from showing up.

`resetFilters` and `resetCaughtDropdown` are harmless, filter-related functions.

`showText` is also harmless.

`setUpgradeAnimation` and `setUpgradeIcon` are used to show which starters have candy upgrades available, I have not touched these at all. There are also `updateCandyUpgradeDisplay` and `onCandyUpgradeDisplayChanged`.

Then we have `processInput`, which calls one of `processFilterModeInput`, `processStartCursorInput`, `processRandomCursorInput`, `processCycleButtonsInput`, `processPartyIconInput`, `processBoxInput`, or `openPokemonMenu`.
The latter is particularly beefy, altogether these functions are over 1000 lines of code. Some of these have been further split up into smaller functions, in particular this was done for `processCycleButtonsInput` to have different functions to set a specific preference. Where applicable, the new functions are also used to update information from the menu.

`isInParty`, `addToParty`, `updatePartyIcon`, and later `popStarter`, `updateStarterValueLabel`, `tryUpdateValue` are all functions related to adding or removing starters to the column on the right.

`switchMoveHandler`, `updateSelectedStarterMoveset` are related to the menu which allows to reorder moves and pick the moveset.

`updateButtonIcon`, `updateFilterButtonIcon`, `updateInstructions` are all related to the various buttons that either cycle through preferences or jump to the filters.

`updateStarters` is a huge one, the function that applies challenges and filters, thereby restricting the list of starters displayed.

`updateScroll` is another big one, also needs to be renamed, and is the one which checks which starters actually need to be displayed on the screen, and updates all the containers.

`setSpecies` is a function which is used to update various data based on the species currently selected in the box. It calls the `.setSpecies` function of the `starterSummary` which updates the sprite, information such as name, groth rate, eggs hatched and so on. There is still some redundancy here which can probably be removed.
I also made a `setNoSpecies` function which clears the information when the cursor goes somewhere else---although this may not be strictly necessary.

`startIconAnimation`, `stopIconAnimation` are utility function to update the bouncing animations of icons.

`setSpeciesDetails` is another big one which has been split up, it is used to update everything that the user can directly change (nature, ability, tera, shiny, etc) and also calls `starterSummary`'s `setSpeciesDetails` function which updates the visual elements. Again, it seems that some stuff is changed in both setSpecies and here which does not need to be.
`resetSpeciesDetails` is called for uncaught mons.
`updateCanCycle` determines whether buttons to cycle nature, shiny etc should be shown.
`updateSpeciesMoves` updates the moveset. It also has a call to `starterSummary` for the visual elements, which is a bit convoluted.

`tryStart` might be moved to another file entirely in the future, if we want to allow restarting a run from stored data. It uses `isPartyValid`.

`toggleStatsMode` is a function called to turn ivs on and off. It can use some clean-up, although it is small.

Then there are a few more utility functions, and some functions to clear data and exit the screen.


## Screenshots/Videos


## How to test the changes?


## Checklist
- The PR content is correctly formatted:
  - [ ] **I'm using `beta` as my base branch**
  - [ ] **The current branch is not named `beta`, `main` or the name of another long-lived feature branch**
  - [ ] I have provided a clear explanation of the changes within the PR description
  - [ ] The PR title matches the Conventional Commits format (as described in [CONTRIBUTING.md](https://github.com/pagefaultgames/pokerogue/blob/beta/CONTRIBUTING.md#pr-title-format))
- [ ] The PR is self-contained and cannot be split into smaller PRs
- [ ] There is no overlap with another open PR
- The PR has been confirmed to work correctly:
  - [ ] I have tested the changes manually
  - [ ] The full automated test suite still passes (use `pnpm test:silent` to test locally)
  - [ ] I have created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes if necessary
- [ ] I have provided screenshots/videos of the changes (if applicable)
  - [ ] I have made sure that any UI changes work for both the default and legacy UI themes (if applicable)

Are there any localization additions or changes? If so:
- [ ] I have created an associated PR on the [locales](https://github.com/pagefaultgames/pokerogue-locales) repository
  - If so, include a link to the PR here: _____
- [ ] I have contacted the Translation Team on Discord for proofreading/translation (contacting the devs in #pokerogue-dev is sufficient, we can then forward the PR to the TL team)

Does this require any additions or changes to in-game assets? If so:
- [ ] I have created an associated PR on the [assets](https://github.com/pagefaultgames/pokerogue-assets) repository
  - If so, include a link to the PR here: _____

Wlowscha committed:
[UI/UX][Refactor] Starter select UI refactor (#6390)

* Moved some functions to utility file

* Moved some process input code to its own functions

* Moving processinput for cycle buttons to its own function

* Moved pokemon menu to its own function

* Replaced setSpecies(null) with setNoSpecies (still some code duplication)

* Hide sprite when moving to random starter button

* Renamed `StarterAttributes` to `StarterPreferences` and related changes

* Cleaned up redundancy; no need to set ivs to `null` since the container is hidden anyway

* Refactoring caughtAttr / seenAttr / no switch in `setSpecies`

* Reducing complexity

* Setting sprite correctly for seen mons

* Extracting logic for seen species from setSpeciesDetails

* Separate setup function for filter bar and instruction buttons

* Moved preferences and other info to their own containers

* Setup function for starter column

* Reduced number of containers; cursor is broken

* Fixed cursor movement

* ??= {}

* Split up statistics from permanent info

* Removed "forSeen" variable

* Function which updates displayed moves

* updateCanCycle function

* set passive details to its own function

* Removed pointless conditionals from setSpeciesDetails

* Added some comments

* Small adjustments

* Splitting up starter summary

* Finished splitting off starter summary (minor bugs)

* Added some comments

* Introduced `getSpeciesPropsFromPreferences` utility function

* Moved some constants and functions around

* Moved ssui to handlers folder

* Also moved some other files

* Formatting

* Fixed wrong input

* Fixed another broken import

Wlowscha committed:
Put new files into place

Wlowscha committed:
Renamed fields in `StarterPreferences`

Wlowscha committed:
Added various functions for setting species details

Wlowscha committed:
Cycling gender now works correctly

fabske0 committed:
[bug] ssui refactor Fix pokerus and select borders (#6517)

fix pokerus and select borders

fabske0 committed:
[bug] ssui refactor show correct tint (#6518)

* fix tint for too expensive pokemons

* fix tint reset on filter

Wlowscha committed:
Merge branch 'beta' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Fixing starter duplication bug, ran biome all

Wlowscha committed:
Shiny and variant also cycle correctly again

Wlowscha committed:
Hiding pokemon info when going to filters

fabske0 committed:
[ui] ssui refactor: show starter cost above sprite (#6519)

* fix cost below sprite

* move to StarterContainer

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Merge branch 'ssui-refactor' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Merge branch 'beta' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Fixed bugs, ran biome

Wlowscha committed:
Merge branch 'beta' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Run biome

DayKev commented on src/ui/containers/starter-container.ts:
```diff
@@ -108,6 +93,39 @@ export class StarterContainer extends Phaser.GameObjects.Container {
+
+    if (this.icon) {
+      this.remove(this.icon);
+      this.icon.destroy(); // Properly removes the sprite from memory
```
```suggestion
      this.icon.destroy();
```

DayKev commented on src/ui/containers/starter-container.ts:
```diff
@@ -108,6 +93,39 @@ export class StarterContainer extends Phaser.GameObjects.Container {
+    }
+
+    // icon
+    this.icon = globalScene.add.sprite(-2, 2, species.getIconAtlasKey(formIndex, shiny, variant));
+    this.icon.setScale(0.5);
+    this.icon.setOrigin(0, 0);
+    this.icon.setFrame(species.getIconId(female, formIndex, shiny, variant));
+    this.checkIconId(female, formIndex, shiny, variant);
+    this.icon.setTint(0);
+    this.add(this.icon);
+    this.icon.setBelow(this.label);
```
```suggestion
    this.icon = globalScene.add
      .sprite(-2, 2, species.getIconAtlasKey(formIndex, shiny, variant))
      .setScale(0.5)
      .setOrigin(0)
      .setFrame(species.getIconId(female, formIndex, shiny, variant))
      .setTint(0)
      .setBelow(this.label);
    this.checkIconId(female, formIndex, shiny, variant);
    this.add(this.icon);
```

Wlowscha committed:
Apply suggestions from code review

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Fixed bug from suggestions

Wlowscha committed:
Removed public folder

Wlowscha committed:
Removed public folder

Wlowscha committed:
Merge branch 'beta' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Fixed redundant methods in pokedex ui

Wlowscha committed:
Fix to tr language missed during merge

Wlowscha committed:
Merge branch 'beta' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Fixed errors from merge conflict

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Merge branch 'ssui-refactor' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Fixes left out of the merge

Wlowscha committed:
Unified usage of `SpeciesDetails`

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Using utils functions in Pokédex; fixed bug that crashed game on opening dex

Wlowscha committed:
`allowTera` in `StarterSummary` now gets updated every time that `setSpecies` is called

Wlowscha committed:
Merge branch 'ssui-refactor' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Fixed broken pokedex test

Wlowscha committed:
Hiding egg moves when they are not defined (Pikachu is back at it)

Wlowscha committed:
Updated several methods to take `StarterSpeciesId` as argument instead of `Species` or `SpeciesId`

Wlowscha committed:
Using `StarterSpeciesId` for list of Pokerus mons

Wlowscha committed:
`checkStarterValidForChallenge` also uses starter id

Wlowscha committed:
Replaced `lastSpecies` with `lastStarterId`

Wlowscha committed:
Removed `SpeciesId` import

Wlowscha committed:
Fixed broken test

Wlowscha committed:
Removed unused input parameter from `getDexAttrProps`

Wlowscha committed:
Updated more functions; starter filters work correctly, including challenges

Wlowscha committed:
Removed log message

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Fixed `getDexAttrProps` call in game over phase

Wlowscha committed:
Made some functions into methods; sound effect plays properly when scrolling up and down

Wlowscha committed:
Splitting cursor update to its own function

Wlowscha committed:
Removed cursors for ability, nature etc.

Wlowscha committed:
Added comments to `setStarterDetails`

Wlowscha committed:
Moved instructions in starter select ui to their own class

Wlowscha committed:
Removed more redundant attributes: `speciesStarterDexEntry` and `speciesStarterMoves`

Wlowscha committed:
Some comments and renaming

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Fixed bug that caused icon to flash when pressing missing cycle buttons

Wlowscha committed:
Removed `getSpeciesDefaultDexAttr`

Wlowscha committed:
Removed some debug logs

Wlowscha committed:
Merge branch 'ssui-refactor' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Updated test file with new name

Wlowscha committed:
Fixing tests broken due to `setBelow`

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

DayKev committed:
Add missing `setBelow` to `mock-sprite.ts`

DayKev committed:
preparation for merge

DayKev committed:
Merge branch 'beta' into ssui-refactor

DayKev committed:
Post-merge cleanup

DayKev committed:
Add visibility modifiers and return types, remove unused `toBoxCursor`

fabske0 commented on src/utils/data.ts:
```diff
@@ -79,19 +79,19 @@ export function isBareObject(obj: any): boolean {
-  [key: number]: StarterAttributes | undefined;
+export interface AllStarterPreferences {
+  [key: number]: StarterPreferences | undefined;
 }
```
```suggestion
export type AllStarterPreferences = {
  [key in SpeciesId]: StarterPreferences | undefined;
}
```

fabske0 commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -4749,16 +3253,16 @@ export class StarterSelectUiHandler extends MessageUiHandler {
    * Designed to be used for unit tests that utilize this UI.
    */
-  clearStarterPreferences() {
+  public clearStarterPreferences(): void {
     this.starterPreferences = {};
     this.originalStarterPreferences = {};
   }
```
```suggestion
  public clearStarterPreferences(): void {
    this.starterPreferences = {} as AllStarterPreferences;
    this.originalStarterPreferences = {} as AllStarterPreferences;
  }
```

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+  gameData ??= globalScene.gameData;
+
+  const starterId = getStarterSpeciesId(speciesId);
+  const hatchCount = globalScene.gameData.dexData[starterId].hatchedCount;
```
Why does this use the `globalScene` one and not the one from the parameter?

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+  const starterId = getStarterSpeciesId(speciesId);
+  gameData ??= globalScene.gameData;
+  const starterData = gameData.starterData[starterId];
+
```
Could be applied to all of these functions
```suggestion
export function isPassiveAvailable(speciesId: number, gameData = globalScene.gameData): boolean {
  // Get this species ID's starter data
  const starterId = getStarterSpeciesId(speciesId);
  const starterData = gameData.starterData[starterId];

```

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+  );
+}
+
+export function isStarterValidForChallenge(starterId: StarterSpeciesId) {
```
```suggestion
/**
 * Determines if a starter is valid for challenges.
 * @param starterId - The ID of the starter species to check
 * @returns whether the starter is valid for challenges
 */
export function isStarterValidForChallenge(starterId: StarterSpeciesId): boolean {
```

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+    allFormsValid = isValidForChallenge;
+  }
+
+  return allFormsValid;
```
```suggestion
  let isStarterValid = false;
  if (species.forms?.length > 0) {
    for (let i = 0; i < species.forms.length; i++) {
      /* Here we are making a fake form index dex props for challenges
       * Since some pokemon rely on forms to be valid (i.e. blaze tauros for fire challenges), we make a fake form and dex props to use in the challenge
       */
      if (!species.forms[i].isStarterSelectable) {
        continue;
      }
      const tempFormProps = BigInt(Math.pow(2, i)) * DexAttr.DEFAULT_FORM;
      const isValidForChallenge = checkStarterValidForChallenge(
        starterId,
        globalScene.gameData.getDexAttrProps(tempFormProps),
        true,
      );
      isStarterValid ||= isValidForChallenge;
    }
  } else {
    const isValidForChallenge = checkStarterValidForChallenge(
      starterId,
      globalScene.gameData.getSpeciesDefaultDexAttrProps(species.speciesId),
      true,
    );
    isStarterValid = isValidForChallenge;
  }

  return isStarterValid;
```

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+
+export function getStarterSelectTextSettings(): StarterSelectLanguageSetting {
+  const currentLanguage = i18next.resolvedLanguage ?? "en";
+  const langSettingKey = Object.keys(languageSettings).find(lang => currentLanguage.includes(lang)) ?? "en";
+  const textSettings = languageSettings[langSettingKey];
+  return textSettings;
```
```suggestion
  const currentLanguage = i18next.resolvedLanguage ?? "en";
  const textSettings = languageSettings[currentLanguage] ?? languageSettings["en"];
  return textSettings;
```

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+  // Unpacking to make a copy by values, not references
+  const copiedDexEntry = { ...dexEntry };
+  copiedDexEntry.ivs = [...dexEntry.ivs];
+  const copiedStarterDataEntry = { ...starterDataEntry };
```
```suggestion
  // Unpacking to make a copy by values, not references
  const copiedDexEntry = deepCopy(dexEntry);
  const copiedStarterDataEntry = { ...starterDataEntry };
```

Wlowscha replied:
Is removing that line intentional?

fabske0 replied:
I changed the first line to use `deepCopy` which should handle the ivs

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+  return { dexEntry: { ...copiedDexEntry }, starterDataEntry: { ...copiedStarterDataEntry } };
+}
+
+export function getFriendship(speciesId: number) {
```
```suggestion
/**
 * Get the current friendship and friendship cap for a given species.
 * @param speciesId - The id of the species to get friendship for
 * @returns An object containing the current friendship and friendship cap for the species
 */
export function getFriendship(speciesId: SpeciesId): { currentFriendship: number; friendshipCap: number } {
```

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+    teraType,
+  };
+}
+
```
```suggestion
/**
 * Get the limit on starter points are available for the current run.
 * @returns the limit on starter points taking challenges into account
 */
```

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+
+  return valueLimit.value;
+}
+
```
```suggestion
/**
 * Calculate the total value of a given party.
 * @param party - An array of species IDs representing the player's starter party
 * @returns The total value of the party
 */
```

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+    0,
+  );
+}
+
```
```suggestion
/**
 * Sort an array of {@linkcode SpeciesId | species IDs} based on a given criteria and direction.
 * @param speciesIds - An array of species IDs to be sorted
 * @param sort - The criteria by which the species hould be sorted
 * @param dir - The direction in which the species should be sorted
 */
```

fabske0 commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,514 @@
+  });
+}
+
+export function getStarterMoves(starterId: StarterSpeciesId, formIndex: number) {
```
```suggestion
/**
 * Get the moves that a starter can have.
 * @param starterId - The id of the starter species to get moves for
 * @param formIndex - The form index of the starter to get moves for
 * @returns An array of move IDs
 */
export function getStarterMoves(starterId: StarterSpeciesId, formIndex: number): MoveId[] {
```

fabske0 reviewed (commented):
Partial

Wlowscha committed:
Merge branch 'ssui-refactor' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Added suggestions from github review

Wlowscha committed:
Update src/ui/utils/starter-select-ui-utils.ts

Co-authored-by: Fabi <192151969+fabske0@users.noreply.github.com>

fabske0 committed:
Merge branch 'beta' of into ssui-refactor

fabske0 committed:
fix errors

fabske0 committed:
update submodules

fabske0 committed:
re-add `StarterSpeciesId`

fabske0 committed:
fix type error

emdeann commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,527 @@
+   *  If neither of these pass, we add DexAttr.MALE to our temp props
+   */
+  if (
+    starterPreferences[speciesId]?.female
```
```suggestion
    starterPreferences.female
```
The passed `StarterPreferences` is already species-specific, it doesn't have a `speciesId` field. Currently selected variant, gender, etc aren't saved when a starter is selected because of this

emdeann commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,527 @@
+   * If they're not there, it enables shiny state by default if any shiny was caught
+   */
+  if (
+    starterPreferences[speciesId]?.shiny
```
```suggestion
    starterPreferences.shiny
```

emdeann commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,527 @@
+    || ((caughtAttr & DexAttr.SHINY) > 0n && starterPreferences[speciesId]?.shiny !== false)
+  ) {
+    props += DexAttr.SHINY;
+    if (starterPreferences[speciesId]?.variant !== undefined) {
```
```suggestion
    if (starterPreferences.variant !== undefined) {
```

emdeann commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,527 @@
+  ) {
+    props += DexAttr.SHINY;
+    if (starterPreferences[speciesId]?.variant !== undefined) {
+      props += BigInt(Math.pow(2, starterPreferences[speciesId]?.variant)) * DexAttr.DEFAULT_VARIANT;
```
```suggestion
      props += BigInt(Math.pow(2, starterPreferences.variant)) * DexAttr.DEFAULT_VARIANT;
```

emdeann commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,527 @@
+    props += DexAttr.NON_SHINY;
+    props += DexAttr.DEFAULT_VARIANT; // we add the default variant here because non shiny versions are listed as default variant
+  }
+  if (starterPreferences[speciesId]?.form) {
```
```suggestion
  if (starterPreferences.form) {
```

emdeann commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,527 @@
+  }
+  if (starterPreferences[speciesId]?.form) {
+    // this checks for the form of the pokemon
+    props += BigInt(Math.pow(2, starterPreferences[speciesId]?.form)) * DexAttr.DEFAULT_FORM;
```
```suggestion
    props += BigInt(Math.pow(2, starterPreferences.form)) * DexAttr.DEFAULT_FORM;
```

emdeann commented on src/ui/utils/starter-select-ui-utils.ts:
```diff
@@ -0,0 +1,527 @@
+
+  // Unpacking to make a copy by values, not references
+  const copiedDexEntry = { ...dexEntry };
+  copiedDexEntry.ivs = { ...dexEntry.ivs };
```
Is this a typo? I don't think we want to spread the ivs into an object

emdeann commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -1550,1235 +832,1304 @@ export class StarterSelectUiHandler extends MessageUiHandler {
+    // Updating tera type for new form
+    this.setNewTeraType(speciesId, getPokemonSpecies(speciesId).forms[formIndex].type1);
+    // Updating gender for gendered forms
+    if (getPokemonSpecies[speciesId]?.forms?.find(f => f.formKey === "female")) {
```
This is indexing into a function

emdeann commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -1550,1235 +832,1304 @@ export class StarterSelectUiHandler extends MessageUiHandler {
+    (this.starterPreferences[speciesId] ??= {}).female = female;
+    (this.originalStarterPreferences[speciesId] ??= {}).female = female;
+    // Updating form for gendered forms
+    if (getPokemonSpecies[speciesId]?.forms?.find(f => f.formKey === "female")) {
```
Same thing here

emdeann commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -3334,1009 +2482,470 @@ export class StarterSelectUiHandler extends MessageUiHandler {
-        }
-        if (variant === undefined || variant !== props.variant) {
-          variant = props.variant;
+        if (caughtAttr & species.getFullUnlocksData() || globalScene.dexForDevs) {
```
This is more strict than it was before, right? I'm noticing weirdness - for example a mon gets grayed out in ssui after purchasing its passive

Wlowscha replied:
I don't see this. You mean the icon gets greyed out?

emdeann replied:
The icon gets grayed out when buying a passive, yes. One of the saves I tested on was the pokédex test save, it happens if you load that and purchase a passive from ssui

emdeann commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -3334,1009 +2482,470 @@ export class StarterSelectUiHandler extends MessageUiHandler {
+      this.setStarterDetails(starterId, false);
 
-        this.setTypeIcons(speciesForm.type1, speciesForm.type2);
+      this.startIconAnimation(this.cursor);
```
When hovering a pokemon in the party column, the last-hovered slot in the grid stays animated. That wasn't the case before, i'm pretty sure it's because this is going by cursor only

emdeann commented on src/ui/containers/starter-container.ts:
```diff
@@ -108,6 +95,40 @@ export class StarterContainer extends Phaser.GameObjects.Container {
+    this.add(this.icon);
+
+    [
+      this.label,
```
This seems like it's missing some stuff? Right now for example the shiny sparkles render below the pokemon icon, where they were above it before

Wlowscha replied:
That's a stylistic choice more than anything, I would say

emdeann replied:
Sure, but it should be explicitly mentioned because it is a functional change

emdeann commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -1550,1235 +832,1304 @@ export class StarterSelectUiHandler extends MessageUiHandler {
-    if (this.isUpgradeAnimationEnabled()) {
-      this.setUpgradeAnimation(starterContainer.icon, this.lastSpecies, true);
+    if (isUpgradeAnimationEnabled()) {
+      this.setUpgradeAnimation(starterContainer.icon, speciesDataRegistry.getSpecies(this.lastStarterId), true);
```
Is there a reason to rely on `lastStarterId` for this, considering that there is a `starterContainer.species` which would seemingly be more reliable?

Wlowscha replied:
Do you mean in this specific instance, or are you arguing against lastStarterId in general?

emdeann replied:
More in general, I'd like to avoid relying on stuff like that as much as possible. But I guess with the changes that I'm working on after this it wouldn't matter much

emdeann reviewed (changes requested):
The only thing I don't have an exact code location for is that the available upgrade bouncing animation doesn't seem to work anymore. It seems like the icon anim manager would be holding onto dead references once a scroll happens. I think in `StarterContainer` you could just `setTexture` on the icon instead of removing and recreating it every time?

Wlowscha committed:
Apply suggestions from code review

Remove obsolete `speciesId` from starter-select-utils.ts

Co-authored-by: Dean <69436131+emdeann@users.noreply.github.com>

Wlowscha committed:
Fixed using `form` instead of `formIndex` in function

Wlowscha committed:
Fixed brackets for ivs

Wlowscha committed:
Fixed indexing into getPokemonSpecies calls

Wlowscha committed:
Icon animations stop when moving off the grid

Wlowscha committed:
Set sprite at the back in icons and summary

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Merge branch 'ssui-refactor' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

fabske0 committed:
add some `readonly`

Wlowscha committed:
Merge branch 'beta' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Removed instances of `getPokemonSpecies`

Wlowscha committed:
Also fixed test

Wlowscha committed:
Merge branch 'beta' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Fix botched import

Wlowscha committed:
PokemonIconAnimHelper now controls the jumping animation in start select

Wlowscha committed:
Added some documentation in the icon animation helper

Wlowscha committed:
Merge branch 'beta' of https://github.com/pagefaultgames/pokerogue into ssui-refactor

Wlowscha committed:
Fixed imports of `AllStarterPreferences`

DayKev committed:
Merge branch 'beta' into ssui-refactor

DayKev commented on src/@types/starter-species-id.ts:
```diff
@@ -0,0 +1,575 @@
+import type { SpeciesId } from "#enums/species-id";
+
+export type StarterSpeciesId =
```
This needs to be auto-generated by a script, with a test to make sure it doesn't get out of sync. Having this need to be manually maintained is not good.

DayKev commented on src/@types/save-data.ts:
```diff
@@ -101,19 +101,22 @@ export interface StarterMoveData {
+  abilityIndex?: number | undefined;
   variant?: number | undefined;
-  form?: number | undefined;
+  formIndex?: number | undefined;
```
The changes to these field names will need migrators (we don't have a dedicated type of migrator for misc localstorage settings, so you can just make a system data migrator and ignore the migrator function's `data` param).

DayKev commented on src/ui/containers/starter-summary.ts:
```diff
@@ -0,0 +1,909 @@
+
+    const species = speciesDataRegistry.getSpecies(starterId);
+
+    // We will only update the sprite if there is a change to form, shiny/variant
+    // or gender for species with gender sprite differences
+    const shouldUpdateSprite =
+      (species.genderDiffs && female != null) || formIndex != null || shiny != null || variant != null;
```
These can never be `null` here anymore due to the changes you've made; this variable will always be `true` now.

DayKev commented on src/ui/containers/starter-summary.ts:
```diff
@@ -0,0 +1,909 @@
+    variant?: Variant | undefined,
+  ): void {
+    species.loadAssets(female, formIndex, shiny, variant, true).then(() => {
+      if (this.assetLoadCancelled?.value) {
```
This can never be `true` any more due to the changes you made. In the old version of SSUI this did not check `this.assetLoadCancelled`, it checked the local `assetLoadCancelled` variable which *could* be `true` at this point due to the absolutely cursed code involved.
So `StarterSummary#assetLoadCancelled` is currently useless, it does nothing.

DayKev committed:
Minor formatting/etc

DayKev committed:
Modify `starter-summary.ts`

Lots of reformatting, de-nesting, simplification,
removal of obsolete comments

DayKev committed:
Reformatting/simplification/doc cleanup in `starter-select-ui-utils.ts`

DayKev committed:
Moves types to `@types/`

DayKev committed:
Fix type of `StarterPreferences#variant`, move enum to `enums/`

DayKev committed:
Remove unused `PokemonIconAnimHelper#remove` method

Remove `PokemonIconAnimHelper#setup` and put
its code into a constructor instead

Remove `PokemonIconAnimHelper#jumpOffset` which never changed,
and simplify `switch` block where it was referenced

DayKev commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -1422,124 +759,39 @@ export class StarterSelectUiHandler extends MessageUiHandler {
+
     // Skip animations if they are disabled
     if (globalScene.candyUpgradeDisplay === 0 || species.speciesId !== species.getRootSpeciesId(false)) {
+      this.iconAnimHandler.addOrUpdate(icon, PokemonIconAnimMode.NONE); // Check that this does not interfere with setting mode to ACTIVE.
```
Is this supposed to be a "todo" comment?

DayKev commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -1553,1235 +805,1310 @@ export class StarterSelectUiHandler extends MessageUiHandler {
+        break;
+      case Button.DOWN:
+        if (currentRow < numOfRows - 1 && this.cursor + 9 < this.filteredStarterIds.length) {
+          // not last row
+          if (currentRow - this.scrollCursor === 8) {
+            // last row of visible starters
```
These comments seem to be giving conflicting messages.

DayKev committed:
Formatting, de-nesting, cleanup of `starter-select-ui-handler.ts`

DayKev committed:
Merge branch 'beta' into ssui-refactor

DayKev commented on src/system/game-data.ts:
```diff
@@ -2048,7 +2048,7 @@ export class GameData {
 
   /**
-   * Converts Pokédex attributes from {@linkcode bigint} to a readable {@linkcode DexAttrProps} interface.
+   * Converts Pokédex attributes from a `bigint` to a readable {@linkcode DexAttrProps} interface.
```
Using a `@linkcode` on a primitive is useless.

DayKev commented on src/ui/containers/starter-summary.ts:
```diff
@@ -521,10 +538,9 @@ export class StarterSummary extends Phaser.GameObjects.Container {
         .setVisible(!!luck)
         .setText(luck.toString())
-        .setTint(getVariantTint(Math.min(luck - 1, 2) as Variant));
+        .setTint(getVariantTint(Phaser.Math.Clamp(luck - 1, 0, 2) as Variant));
```
It's possible for `luck - 1` to be a negative number, which is an invalid input for `getVariantTint()`.

DayKev reviewed (changes requested):
[Review changes](https://github.com/pagefaultgames/pokerogue/pull/6516/changes/05c4b560b8dd241ad289ce3a3056ea3a411229a9..2168fc607065224aded1e23cc1e8e6941e6f9689)

---

The Pokémon minisprite in the current party box on the right displays as shiny if you add a non-shiny to the party, only updating to the correct non-shiny minisprite when you move the cursor onto it (or the original in the starter grid). This also causes you to start the game with a shiny Pokémon if you start while the minisprite is incorrect.
If the Pokémon is shiny when you add it, the minisprite is correct from the start.
This only occurs if you have the shiny unlocked for the starter, when testing on a fresh account via private browsing it doesn't incorrectly make starters shiny.

<details><summary>Screenshot</summary>

<img width="804" height="357" alt="image" src="https://github.com/user-attachments/assets/79b90911-e20b-4761-a9cc-e460f9c81dfe" />

</details>

DayKev committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Merge branch 'beta' into ssui-refactor

Wlowscha committed:
Improved some comments

Wlowscha committed:
Added a TODO

Wlowscha committed:
Removed assetLoadCancelled

Wlowscha committed:
Removed shouldUpdateSprite from setStarterDetails

Wlowscha committed:
Fixed wrong access to `starterPreferences`

Wlowscha committed:
Migrator to correctly change key names in preferences in local storage

Wlowscha committed:
Added test to check that all `StarterSpeciesId` have a corresponding species in the register

Wlowscha committed:
Added Species data registry test file to exclude rule by depcruise

Wlowscha committed:
Fixed migrator to correct version

Wlowscha committed:
Merge branch 'beta' of https://github.com/pagefaultgames/pokerogue into ssui-refactor
